lib/Driver/ToolChains.cpp

```diff
@@ -526,6 +401,185 @@ ToolChain::constructInvocation(const CompileJobAction &job,
+    Arguments.push_back(getAllSourcesPath());
+  }
+  if (UsePrimaryFileList) {
+    Arguments.push_back("-primary-filelist");
+    Arguments.push_back(getTemporaryFilePath("primaryInputs", ""));
+    FilelistInfos.push_back({Arguments.back(), types::TY_Swift,
+                             FilelistInfo::WhichFiles::PrimaryInputs});
+  }
+  if (!UseFileList || !UsePrimaryFileList) {
+    addFrontendInputArguments(MayHavePrimaryInputs, UseFileList,
+                              UsePrimaryFileList, FilterInputsByType,
+                              Arguments);
+  }
+
+  if (UseSupplementaryOutputFileList) {
+    Arguments.push_back("-supplementary-output-file-map");
+    Arguments.push_back(getTemporaryFilePath("supplementaryOutputs", ""));
+    FilelistInfos.push_back({Arguments.back(), types::TY_INVALID,
+                             FilelistInfo::WhichFiles::SupplementaryOutput});
+  } else {
+    addFrontendSupplementaryOutputArguments(Arguments);
+  }
+}
+
+void ToolChain::JobContext::addFrontendInputArguments(
+    const bool mayHavePrimaryInputs, const bool useFileList,
+    const bool usePrimaryFileList, const bool filterByType,
+    ArgStringList &arguments) const {
+  llvm::StringSet<> primaries;
+
+  if (mayHavePrimaryInputs)
+    for (const Action *A : InputActions) {
+      const auto *IA = cast<InputAction>(A);
+      const llvm::opt::Arg &InArg = IA->getInputArg();
+      // For an index data job,
+      // only make the index-file-path ones primary
+      if (ToolChain::canCompileInputArgumentBePrimary(Output, InArg))
+        primaries.insert(InArg.getValue());
+    }
+  for (auto inputPair : getTopLevelInputFiles()) {
+    if (filterByType && !types::isPartOfSwiftCompilation(inputPair.first))
+      continue;
+    const char *inputName = inputPair.second->getValue();
+    if (primaries.count(inputName)) {
+      if (!usePrimaryFileList) {
+        arguments.push_back("-primary-file");
+        arguments.push_back(inputName);
+      }
+    } else {
+      if (!useFileList) {
+        arguments.push_back(inputName);
+      }
+    }
+  }
+}
+
+bool ToolChain::canCompileInputArgumentBePrimary(const CommandOutput &Output,
+                                                 const llvm::opt::Arg &A) {
+  return Output.getPrimaryOutputType() != types::TY_IndexData ||
+         A.getOption().getID() == options::OPT_index_file_path;
+}
+
+void ToolChain::JobContext::addFrontendSupplementaryOutputArguments(
+    ArgStringList &arguments) const {
+  // FIXME: get these and other argument strings from the same place for both
+  // driver and frontend
+  addOutputsOfType(arguments, Output, Args, types::ID::TY_SwiftModuleFile,
+                   "-emit-module-path");
+
+  addModuleDocPathArgIfNecessary(Output, Args, arguments);
+
+  addSerializeDiagnosticsPathArgIfNecessary(Output, Args, arguments);
+
+  if (addOutputsOfType(arguments, Output, Args, types::ID::TY_ObjCHeader,
+                       "-emit-objc-header-path")) {
```

Comment:
Why did you factor out `addModuleDocPathArgIfNecessary` but not `-emit-objc-header-path`?

Reply:
`-emit-objc-header-path` is only created in `ToolChain::constructInvocation(const MergeModuleJobAction &job, const JobContext &context)`.

Reply:
Thanks, great reminder!

Reply:
That can't be right. It's needed for SingleCompile too.

Reply:
Sorry, I mean "it's created here, for SingleCompile, and also in MergeModuleJobAction".

Reply:
I am so glad you pointed this out! When I looked at it, I decided to make two changes:
1. Eliminate a level of (over) factoring. I decided that `addModuleDocPathArgIfNecessary` and its brethren were over-egging the pudding.
2. Add a bit of symmetry: compile jobs were emitting the objc header path via `addOutputsOfType` but the merge-module path was not. I changed the latter path to be symmetric with the former.

I'm hoping this arrangement will feel better.